filters/filter.go

```diff
@@ -56,32 +71,38 @@ type Filter interface {
 }
 
 // NewFilter supports startsWith, endsWith, contains and a single wildcard
-// along with negation
-// TODO(martinm): add rest of glob matching support
+// along with negation and glob matching support
+// TODO(martinm): Add variable length glob ranges
 func NewFilter(pattern string) (Filter, error) {
-	if len(pattern) == 0 || pattern[0] != negationChar {
-		return newFilter(pattern)
+	if len(pattern) == 0 {
+		return newEqualityFilter(pattern), nil
+	}
+
+	if pattern[0] != negationChar {
+		return newWildcardFilter(pattern)
 	}
 
 	if len(pattern) == 1 {
 		// only negation symbol
 		return nil, errInvalidFilterPattern
 	}
 
-	filter, err := newFilter(pattern[1:])
+	filter, err := newWildcardFilter(pattern[1:])
 	if err != nil {
 		return nil, err
 	}
 
 	return newNegationFilter(filter), nil
 }
 
-func newFilter(pattern string) (Filter, error) {
+// newWildcardFilter creates a filter that segments the pattern based
+// on wildcards, creating a rangeFilter for each segment
+func newWildcardFilter(pattern string) (Filter, error) {
 	wIdx := strings.Index(pattern, wildcardChar)
 
 	if wIdx == -1 {
 		// No wildcards
-		return newEqualityFilter(pattern), nil
+		return newRangeFilter(pattern, false, middle)
```

Comment:
Difference between start and middle is that we expect chars to be left over after the chain of matches for a start, but expect no chars to be left over after a chain of matches for the middle.